### PR TITLE
Support for Bidder Alias GVL IDs

### PR DESCRIPTION
**Pull request by Pubmatic-Dhruv-Sonone**
This pr is to add support for Bidder Alias GVL IDs for GDPR flow.

### CODE REVIEW

**Comment by bsardo** [comment]
Hi @Pubmatic-Dhruv-Sonone. Sorry for the delay. I'll start looking at this today and get back to you shortly.

File: gdpr/impl_test.go [hunk 1]
```diff
@@ -225,14 +230,26 @@ func TestAllowActivities(t *testing.T) {
...
+		{
+			description:      "Allow PI - known Alias vendor GVLID with Yes GDPR",
+			bidderName:       aliasedBidderAllowedByConsent,
+			bidderCoreName:   bidderAllowedByConsent,
+			gdprDefaultValue: "1",
+			gdpr:             SignalYes,
+			consent:          vendor2AndPurpose2Consent,
+			passID:           true,
+			aliasGVLIDs:      map[string]uint16{"appnexus1": 2},
+		},
```
Comment (bsardo) [inline gdpr/impl_test.go:248]: How about adding another test case just below this new one that ensures the GVL for the alias is what's affecting `passID`:
```
{
	description:      "Don't allow PI - known alias vendor with Yes GDPR, alias vendor does not consent to purpose 2",
	bidderName:       aliasedBidderAllowedByConsent,
	bidderCoreName:   bidderAllowedByConsent,
	gdprDefaultValue: "1",
	gdpr:             SignalYes,
	consent:          vendor2AndPurpose2Consent,
	passID:           false,
	aliasGVLIDs:      map[string]uint16{"appnexus1": 1},
},
```
In this case the GVL ID of 1 does not exist in the fetched vendor list so legal basis will not be established resulting in a `passID` of false.

File: gdpr/impl_test.go [hunk 2]
```diff
@@ -453,43 +479,57 @@ func TestAllowActivitiesGeoAndID(t *testing.T) {
...
 		},
 		{
-			description: "Rubicon vendor test, Specific purposes/LIs claimed, no geo claimed",
-			bidder:      openrtb_ext.BidderRubicon,
-			consent:     "COzTVhaOzTVhaGvAAAENAiCIAP_AAH_AAAAAAEEUACCKAAA",
-			allowBid:    true,
-			passGeo:     false,
-			passID:      true,
+			description:    "Pubmatic Alias vendor test, flex purposes claimed",
+			bidder:         "pumatic1",
```
Comment (bsardo) [inline gdpr/impl_test.go:500]: There's a typo here so the alias won't be found in the GVL ID map. I suggest fixing the typo and then adding another test case where the alias is found in the GVL ID map but results in a different outcome than if the core bidder GVL ID was used. Something similar to my suggestion above for `TestAllowActivities` on line 248.

**Review by bsardo — COMMENTED** [review]
@Pubmatic-Dhruv-Sonone Do you want to mark this as ready for review? I already reviewed it and have just a few comments.

File: exchange/utils.go [hunk 3]
```diff
@@ -649,6 +654,19 @@ func parseAliases(orig *openrtb2.BidRequest) (map[string]string, []error) {
 	return aliases, nil
 }
 
+// parseAliasesGVLIDs parses the Bidder Alias GVLIDs from the BidRequest
+func parseAliasesGVLIDs(orig *openrtb2.BidRequest) (map[string]uint16, []error) {
```
[the remark this answers is not in the document]
Comment (Pubmatic-Dhruv-Sonone) [reply exchange/utils.go:658]: Added the unit test.

File: gdpr/impl_test.go [hunk 4]
```diff
@@ -710,27 +765,43 @@ func TestAllowActivitiesBidRequests(t *testing.T) {
...
+			purpose2EnforcePurpose: config.TCF2FullEnforcement,
+			purpose2EnforceVendors: true,
+			bidder:                 openrtb_ext.BidderPubmatic,
+			bidderCoreName:         openrtb_ext.BidderPubmatic,
+			consent:                purpose2ConsentWithoutVendorConsent,
+			allowBid:               false,
+			passGeo:                false,
+			passID:                 false,
+		},
+
```
[the remark this answers is not in the document]
Comment (Pubmatic-Dhruv-Sonone) [reply gdpr/impl_test.go]: Done

File: gdpr/impl_test.go [hunk 5]
```diff
@@ -846,11 +922,14 @@ func TestAllowActivitiesVendorException(t *testing.T) {
 		allowBid              bool
 		passGeo               bool
 		passID                bool
+		bidderCoreName        openrtb_ext.BidderName
+		aliasGVLIDs           map[string]uint16
```
[the remark this answers is not in the document]
Comment (Pubmatic-Dhruv-Sonone) [reply gdpr/impl_test.go]: Passed empty aliasGVLIDs to function

File: gdpr/impl_test.go (hunk 1 above)
Comment (Pubmatic-Dhruv-Sonone) [reply gdpr/impl_test.go:248]: Added the another test case

File: gdpr/impl_test.go (hunk 2 above)
Comment (Pubmatic-Dhruv-Sonone) [reply gdpr/impl_test.go]: Done

**Comment by Pubmatic-Dhruv-Sonone** [comment]
@bsardo  Addressed the review comments. Please check

File: gdpr/gdpr.go [hunk 6]
```diff
@@ -24,7 +24,7 @@ type Permissions interface {
 	// Determines whether or not to send PI information to a bidder, or mask it out.
 	//
 	// If the consent string was nonsensical, the returned error will be an ErrorMalformedConsent.
-	AuctionActivitiesAllowed(ctx context.Context, bidder openrtb_ext.BidderName, PublisherID string, gdprSignal Signal, consent string, weakVendorEnforcement bool) (allowBidReq bool, passGeo bool, passID bool, err error)
+	AuctionActivitiesAllowed(ctx context.Context, bidderCoreName openrtb_ext.BidderName, bidder openrtb_ext.BidderName, PublisherID string, gdprSignal Signal, consent string, weakVendorEnforcement bool, aliasGVLIDs map[string]uint16) (allowBidReq bool, passGeo bool, passID bool, err error)
```
Comment (VeronikaSolovei9) [inline gdpr/gdpr.go:27]: Just for my knowledge. What is the difference between `bidderCoreName` and `bidder`? Why do you need to use `bidderCoreName` here?

**Review by VeronikaSolovei9 — COMMENTED** [review]
Overall it looks good to me, added couple of minor comments

File: endpoints/openrtb2/auction.go [hunk 7]
```diff
@@ -1231,6 +1240,20 @@ func (deps *endpointDeps) validateAliases(aliases map[string]string) error {
...
 
+func (deps *endpointDeps) validateAliasesGVLIds(aliasesGVLIDs map[string]uint16, aliases map[string]string) error {
+	for alias, vendorId := range aliasesGVLIDs {
+
+		if _, aliasExist := aliases[alias]; !aliasExist {
+			return fmt.Errorf("request.ext.prebid.aliasgvlids. vendorId %d refers to unknown bidder alias: %s", vendorId, alias)
+		}
+
+		if vendorId < 1 {
+			return fmt.Errorf("request.ext.prebid.aliasgvlids. Invalid vendorId %d for alias: %s. Choose a different vendorId, or remove this entry.", vendorId, alias)
```
[the remark this answers is not in the document]
Comment (Pubmatic-Dhruv-Sonone) [reply endpoints/openrtb2/auction.go:1247]: Covered this branch in test coverage

File: gdpr/gdpr.go (hunk 6 above)
Comment (Pubmatic-Dhruv-Sonone) [reply gdpr/gdpr.go:27]: bidder is the alias name and bidderCoreName is the actual bidder. Ex: bidder = `brightroll ` and bidderCoreName = `appnexus`. 

bidder is needed to get vendorId from AliasGVLIDs map and bidderCoreName is needed to get vendorId from bidderInfo

File: gdpr/impl_test.go [hunk 8]
```diff
@@ -710,27 +785,44 @@ func TestAllowActivitiesBidRequests(t *testing.T) {
...
+			purpose2EnforceVendors: true,
+			bidder:                 "pubmatic1",
+			bidderCoreName:         openrtb_ext.BidderPubmatic,
+			consent:                purpose2ConsentWithoutVendorConsent,
+			allowBid:               false,
+			passGeo:                false,
+			passID:                 false,
+			aliasGVLIDs:            map[string]uint16{"pubmatic1": 6},
+		},
+
```
Comment (bsardo) [inline gdpr/impl_test.go:819]: How about replacing this new test case with these to prove that the alias GVL ID has a direct effect on whether the bid is allowed:
```
{
			description:            "Bid allowed - p2 enabled, user consents to p2 and vendor, alias vendor consents to p2",
			purpose2EnforcePurpose: config.TCF2FullEnforcement,
			purpose2EnforceVendors: true,
			bidder:                 "pubmatic1",
			bidderCoreName:         openrtb_ext.BidderPubmatic,
			consent:                purpose2AndVendorConsent,
			allowBid:               true,
			passGeo:                false,
			passID:                 true,
			aliasGVLIDs:            map[string]uint16{"pubmatic1": 6},
		},
		{
			description:            "Bid blocked - p2 enabled, user consents to p2 and vendor, alias vendor does not consent to p2",
			purpose2EnforcePurpose: config.TCF2FullEnforcement,
			purpose2EnforceVendors: true,
			bidder:                 "pubmatic1",
			bidderCoreName:         openrtb_ext.BidderPubmatic,
			consent:                purpose2AndVendorConsent,
			allowBid:               false,
			passGeo:                false,
			passID:                 false,
			aliasGVLIDs:            map[string]uint16{"pubmatic1": 1},
		},
```

**Review by bsardo — COMMENTED** [review]
Looks good, just a few minor things.

**Comment by VeronikaSolovei9** [comment]
Thank you for updates. I'm ready to approve it when @bsardo changes are addressed.

File: gdpr/gdpr.go (hunk 6 above)
Comment (VeronikaSolovei9) [reply gdpr/gdpr.go:27]: Thank you for the explanation!

File: endpoints/openrtb2/auction.go [hunk 9]
```diff
@@ -1231,6 +1240,20 @@ func (deps *endpointDeps) validateAliases(aliases map[string]string) error {
 	return nil
 }
 
+func (deps *endpointDeps) validateAliasesGVLIds(aliasesGVLIDs map[string]uint16, aliases map[string]string) error {
```
[the remark this answers is not in the document]
Comment (Pubmatic-Dhruv-Sonone) [reply endpoints/openrtb2/auction.go]: Done

File: exchange/utils_test.go [hunk 10]
```diff
@@ -2881,3 +2881,55 @@ func TestApplyFPD(t *testing.T) {
...
+			},
+			nil,
+			true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, got1 := parseAliasesGVLIDs(tt.args.orig)
+			assert.Equal(t, tt.want, got, "parseAliasesGVLIDs() got = %v, want %v", got, tt.want)
+			if (got1 != nil) != tt.wantError {
```
[the remark this answers is not in the document]
Comment (Pubmatic-Dhruv-Sonone) [reply exchange/utils_test.go]: Changed to` if !tt.wantError && err != nil {`

File: exchange/utils_test.go [hunk 11]
```diff
@@ -2881,3 +2881,55 @@ func TestApplyFPD(t *testing.T) {
...
+		{
+			"Invalid AliasGVLID",
+			args{
+				orig: &openrtb2.BidRequest{
+					Ext: json.RawMessage(`{"prebid":{"aliases":{"brightroll":"appnexus"}, "aliasgvlids":"abc"}`),
+				},
+			},
+			nil,
+			true,
+		},
```
[the remark this answers is not in the document]
Comment (Pubmatic-Dhruv-Sonone) [reply exchange/utils_test.go:2924]: Added test case for missing AliasGVLID

File: gdpr/impl_test.go (hunk 8 above)
Comment (Pubmatic-Dhruv-Sonone) [reply gdpr/impl_test.go]: Replaced the existing test case with these two test cases.

**Review by bsardo — APPROVED** [review]
LGTM

**Review by VeronikaSolovei9 — APPROVED** [review]
LGTM

**Pull request merged**